rfcs/0000-type-stability.md

```diff
@@ -0,0 +1,500 @@
+
+## Detailed design
+
+TypeScript should be treated with exactly the same rigor as other elements of the Ember ecosystem, with the same level of commitment to stability, clear and accurate use of Semantic Versioning, testing, and clear policies about breaking changes.
+
+TypeScript introduces two new concerns around breaking changes for addons which publish types.
+
+1.  TypeScript does not adhere to the same norms around Semantic Versioning as the rest of the npm ecosystem, so it is important for addon authors to understand when TypeScript versions may introduce breaking changes without any other change made to the addon.
+
+2.  The runtime behavior of the addon is no longer the only source of potentially-breaking changes: types may be as well. In a well-typed addon, runtime behavior and types should be well-aligned, but it is possible to introduce breaking changes to types without changing runtime behavior.
+
+Accordingly, we must define breaking changes precisely and carefully.
+
+### Background: TypeScript and Semantic Versioning
+
+TypeScript ***does not*** adhere to Semantic Versioning, but since it participates in the npm ecosystem, it uses the same format for its version numbers: `<major>.<minor>.<patch>`. In Semantic Versioning, `<major>` would be a breaking change release, `<minor>` would be a backwards-compatible feature-addition release, and `<patch>` would be a “bug fix” release.
+
+TypeScript maintains this final distinction: the `<patch>` position corresponds to bug fix changes released between the time of one `<minor>` release and the next. For example, `3.9.3` is the third patch/bug fix release for TypeScript version 3.9. However, both `<major>` and `<minor>` releases *may* introduce breaking changes of the sort that the npm ecosystem normally reserves for the `<major>` slot in the version number. (Not all `<minor>` *or* `<major>` releases *do* introduce breaking changes in the normal Semantic Versioning sense, but either *may*.) Accordingly, and for simplicity, the Ember ecosystem should treat *all* TypeScript `<major>.<minor>` releases as *potentially* representing breaking changes.
+
+By example:
+
+-   `3.9.2` to `3.9.3`: always considered a bug fix
+-   `3.9.3` to `4.0.0`: *may or may not* introduce breaking changes
+-   `3.8.3` to `3.9.0`: *may or may not* introduce breaking changes
+
+### Defining breaking changes
+
+When making a change to the types of the public interface, it is subject to the same constraints as runtime code: *breaking the published types entails a breaking change.* Not all changes to published types are *breaking*, however: some changes will continue to allow user code to continue working without any issue. Thus, a *breaking* change to types is a change where previously-allowed code will stop type-checking.
+
+#### Breaking changes to type definitions
```

Comment:
We're missing a few here

- renaming of interfaces can break user-land code
- reorganization of modules can break user-land code
- changing the module where an interface is _actually defined_ can break user-land code (IIRC interface augmentations _can only be applied to the declaration site_ -- they don't work nicely with re-exports through other modules)
- changing a type alias to an interface can break user-land code

Reply:
Thank you for flagging these up. I could have sworn I had them, actually; but it appears I dropped it in Git somewhere, unfortunately. I believe that re-exports of interfaces are not an issue; I just tested it. It's entirely possible that it *used* to be an issue at some point.

Reply:
Added a description of each of these, as well as a note on removing public symbols and a note on when *adding* public symbols can be breaking.